### ISSUE TITLE
GitHub bot for finding issues in existing repos and creating PRs or issues for them

### ISSUE THREAD

**Issue by KOLANICH** [open-source-ideas/open-source-ideas#63]
## Project description
GitHub search is quite a mighty thing. Creating a right query you can find projects linked to some other project somehow.

Lots of repos have some issues (which are not issues fro mhe point if view of their authors).

For example python repos not using `setup.cfg` and storing all the metadata in `setup.py`. It's bad: in general case you cannot extract metadata from it without executing it. Though in most cases, where the large part of it is just a dict passed into `setup` it may be possible. Storing metadata entirely in `setup.cfg` brings 2 profits:
1 tools can easily parse it
2 setup.py usually doesn't changes from project to project. So it may be possible to have a single boilerplate for it.

So we need a bot scanning github for python repos not using `setup.cfg` for storing metadata, trying to parse metadata from `setup.py` without executing it, adding them into `setup.cfg`, patching  `setup.py` and sending a PR.
If something fails, just creating an issue  there asking owners to migrate.

Another trouble is that lot of projects storing binaries in repos don't use git-lfs for that. This results in a very large size of the repos and slow fetches.

So a bot needed locating such repos and adding needed lines into `.gitattributes` and sending a PR.

Another trouble is that some projects don't use .editorconfig. This makes some troubles to me personally because I personally use tabs and I don't want to change my editor's settings permanently and I'm sure that my PRs won't ge merged if I used tabs or instead of predetermined amount of spaces. If .editorconfig file is present the settings are changed for each file based on its path and the config on the file.

So a bot is needed estimating project's coding conventions from source code, creating .editorconfig and sending a PR.

All these have some stages in common:
1 Interaction with GH API v4 (GraphQL)
2 searching code, repos, issues, etc with user-provided query
3 transforming code links into repos links
4 checking the Bloom filter for already visited repos
5 deduplication
6 user-programmed stage creating the issue/PR internal representation
7 checking if the issue already addressed (searching a PR/issue with similar changes), some machine learning may be involved
8 creating an issue/PR if it's not
9 adding the project name into Bloom Filter not to process it again

**Comment by anishkny** [comment]
> 2 searching code, repos, issues, etc with user-provided query

[BigQuery GitHub dataset] could be useful for this

**Comment by sladyn98** [comment]
@KOLANICH Sounds interesting, sounds like dependabot ?
Is this project idea still viable ?

**Comment by KOLANICH** [comment]
@sladyn98, not quite like dependabot, AFAIK dependabot is about dependencies being outdated. This proposal is about best practices. I mean there are lot of repos that are inconvenient to use because they lack something, that can be added automatically. For example `.editorconfig` files. Or for example lot of repos use `setuptools`, but store the metadata in `setup.py` when at least some of it can be moved to `setup.cfg` making `setup.py` more easily auditable. Or, for example lack of `pyproject.toml`, that prevents usage of `python3 -m pep517.build`. All of this stuff can often (but not always) be derived automatically. Currently when I have to deal with a repo, I usually send there a PR fixing the issues, because I hate mess. Currently I do it semi-automatically, but not fully, it's a bit inefficient. I want a tool doing all this work for me. Also it is a pain in the ass to persuade repos maintainers to do anything. For every change with obvious benefits there are a lot of people for whom the benefits are not obvious, and there are lots of people who don't care of any benefits and take decisions based on "because I just like the way it is done currently". It is very frustrating to negotiate to stubborn people, but I have to, otherwise my labour will be wasted. Also some people just ignore PRs and issues for years. Bots don't negotiate to people, instead they can go to everyone almost in no cost, so if they don't care - it'd be these peoples' problem, not mine, because bots' contributions require no my labour, so if PR is ignored or rejected, almost no of my labour is wasted.

So the issue is real, feel free to work on it.

**Comment by sladyn98** [comment]
@KOLANICH Yeah got it, Some of the features that I can see would need to be addressed
a) Would the bot be run by the user on a specific set of repos or would it just scan github for all python repositories, because there are a huge number of python repositories on github.

>  searching code, repos, issues, etc with user-provided query
Would we want the user to provide the repository ?? This makes sense because it would then open the PR's in only those repositories the user wants to update, but it would then miss out on a large number of repositories that face this problem.
Another solution is to scan a set of repositories every day (say 100) and make changes to those repositories, thereby helping out a large number of people.

**Comment by KOLANICH** [comment]
>Would the bot be run by the user on a specific set of repos or would it just scan github for all python repositories, because there are a huge number of python repositories on github.

it is meant to just scan.  with maybe some optimizations. Basically search GH using its API for repos matching criteria to implemented fixes, then marking them as found, then trying to fix the issues in them (without downloading the full repos, we usually need to fetch a few files), then sending PRs via API, 6hen marking the repos as already visited.

>Would we want the user to provide the repository ??

maybe yes, for dealing to repos in manual mode.

>thereby helping out a large number of people.

We are not quite helping people. People usually don't want these changes. If they wanted, they would have been already implemented without our PRs. We are helping ourselves by transforming the ecosystem repo-by-repo to the state we like to see it. Our goal is to make people consider these conveniences de-facto standards. Once that is happenned, it would be no longer needed to robosend PRs, repos owners will implement the fixes themselves without our "advice".

**Comment by sladyn98** [comment]
@KOLANICH I could begin with a basic Proof of concept using golang, I will build something initial and share it with you :)

**Comment by KOLANICH** [comment]
Thanks, just please keep in mind I am mainly interested in python repos, and ... tools for python are mainly written in python. So we need an interface. I guess shared memory pages with ring buffers to pass messages. At least following message types are needed:
* "I have checked out the repo, please do the fixes if you can"
* "I have done some fixes, please commit them (here is the serialized human-readable descr of fixes) and send a PR"
* "no fixes are available for this repo"
